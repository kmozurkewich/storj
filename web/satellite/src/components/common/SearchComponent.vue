// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

<template>
    <input
        @mouseenter="onMouseEnter"
        @mouseleave="onMouseLeave"
        @input="onInput"
        v-model="searchQuery"
        :placeholder="`Search ${placeHolder}`"
        :style="customWidth"
        type="text">
</template>

<script lang="ts">
    import { Component, Prop, Vue } from 'vue-property-decorator';

    declare type searchCallback = (search: string) => Promise<void>;

    @Component
    export default class SearchComponent extends Vue {
        @Prop({default: ''})
        private readonly placeHolder: string;
        @Prop({default: () => { return ''; }})
        private readonly search: searchCallback;

        private inputWidth: string = '56px';
        private searchQuery: string = '';

        public onMouseEnter(): void {
            this.inputWidth = '602px';
        }

        public onMouseLeave(): void {
            if (!this.searchQuery) {
                this.inputWidth = '56px';
            }
        }

        public onInput(): any {
            this.onMouseLeave();
            this.processSearchQuery();
        }

        public get customWidth(): object {
            return { width: this.inputWidth };
        }

        public async processSearchQuery() {
            await this.search(this.searchQuery);
        }
    }
</script>

<style scoped lang="scss">
    input {
        position: absolute;
        right: 0;
        padding: 0 38px 0 18px;
        border: 1px solid #F2F2F2;
        box-sizing: border-box;
        box-shadow: 0 4px 4px rgba(231, 232, 238, 0.6);
        outline: none;
        border-radius: 36px;
        height: 56px;
        font-family: 'font_regular';
        font-size: 16px;
        transition: all 0.4s ease-in-out;
        background-image: url('../../../static/images/team/searchIcon.svg');
        background-repeat: no-repeat;
        background-size: 22px 22px;
        background-position: top 16px right 16px;
    }

    ::-webkit-input-placeholder {
        color: #AFB7C1;
    }
</style>
