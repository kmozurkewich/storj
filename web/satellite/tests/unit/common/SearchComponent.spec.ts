// Copyright (C) 2019 Storj Labs, Inc.
// See LICENSE for copying information.

import { mount, shallowMount } from '@vue/test-utils';
import * as sinon from 'sinon';
import SearchComponent from '@/components/common/SearchComponent.vue';

describe('SearchComponent.vue', () => {
    it('renders correctly', () => {
        const wrapper = shallowMount(SearchComponent);

        expect(wrapper).toMatchSnapshot();
    });

    it('renders correctly with default props', () => {
        const wrapper = shallowMount(SearchComponent);

        expect(wrapper.find('input').props('placeHolder')).toMatch('');
        expect(wrapper.find('input').props('search')).toMatch('');
    });

    it('functions onMouseEnter/onMouseLeave work correctly', () => {
        const wrapper = mount(SearchComponent);

        wrapper.vm.onMouseEnter();

        expect(wrapper.vm.inputWidth).toMatch('602px');

        wrapper.vm.onMouseLeave();

        expect(wrapper.vm.searchQuery).toMatch('');
        expect(wrapper.vm.inputWidth).toMatch('56px');
    });

    it('function onInput works correctly', () => {
        let onMouseLeaveSpy = sinon.spy();
        let processSearchQuerySpy = sinon.spy();

        const wrapper = mount(SearchComponent, {
            methods: {
                onMouseLeave: onMouseLeaveSpy,
                processSearchQuery: processSearchQuerySpy,
            }
        });

        wrapper.vm.onInput();

        expect(onMouseLeaveSpy.callCount).toBe(1);
        expect(processSearchQuerySpy.callCount).toBe(1);
    });

    it('function processSearchQuery works correctly', async () => {
        const wrapper = mount(SearchComponent);

        wrapper.vm.processSearchQuery();

        expect(wrapper.vm.search).resolves;
    });
});
