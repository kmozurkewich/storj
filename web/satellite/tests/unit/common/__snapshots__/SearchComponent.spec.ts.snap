// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchComponent.vue renders correctly 1`] = `<input placeholder="Search " type="text" style="width: 56px;">`;
